{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jyjiang/ls/envs/drug/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-04-11 14:47:30,243] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to cuda (auto detect)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: cannot find -laio: No such file or directory\n",
      "collect2: error: ld returned 1 exit status\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: warning: librt.so.1, needed by /usr/local/cuda/lib64/libcufile.so, not found (try using -rpath or -rpath-link)\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: warning: libpthread.so.0, needed by /usr/local/cuda/lib64/libcufile.so, not found (try using -rpath or -rpath-link)\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: warning: libstdc++.so.6, needed by /usr/local/cuda/lib64/libcufile.so, not found (try using -rpath or -rpath-link)\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: warning: libm.so.6, needed by /usr/local/cuda/lib64/libcufile.so, not found (try using -rpath or -rpath-link)\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::runtime_error::~runtime_error()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__gxx_personality_v0@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream::tellp()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::chrono::_V2::steady_clock::now()@GLIBCXX_3.4.19'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::_M_replace_aux(unsigned long, unsigned long, unsigned long, char)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for bool@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__throw_logic_error(char const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::logic_error@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::locale::~locale()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::string const&, unsigned long, unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_end_catch@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_ofstream<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::logic_error::~logic_error()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for __cxxabiv1::__si_class_type_info@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ios<char, std::char_traits<char> >::_M_cache_locale(std::locale const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `operator new[](unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::_M_leak_hard()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_ifstream<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_streambuf<wchar_t, std::char_traits<wchar_t> >::basic_streambuf(std::basic_streambuf<wchar_t, std::char_traits<wchar_t> > const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::append(char const*, unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::string const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for unsigned short@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::resize(unsigned long, char)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for char const*@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ctype<char>::_M_widen_init() const@GLIBCXX_3.4.11'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__throw_invalid_argument(char const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::locale::operator=(std::locale const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ios<wchar_t, std::char_traits<wchar_t> >::_M_cache_locale(std::locale const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::_Rb_tree_decrement(std::_Rb_tree_node_base const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_free_exception@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::condition_variable::notify_one()@GLIBCXX_3.4.11'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ios_base::Init::~Init()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_pure_virtual@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream::flush()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for __cxxabiv1::__class_type_info@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_rethrow@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_fstream<char, std::char_traits<char> >::~basic_fstream()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::compare(char const*) const@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_ostringstream<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::locale::locale()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::chrono::_V2::system_clock::now()@GLIBCXX_3.4.19'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_ifstream<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::_Hash_bytes(void const*, unsigned long, unsigned long)@CXXABI_1.3.5'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream& std::ostream::_M_insert<long long>(long long)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for char*@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__detail::_Prime_rehash_policy::_M_need_rehash(unsigned long, unsigned long, unsigned long) const@GLIBCXX_3.4.18'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::out_of_range@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream& std::ostream::_M_insert<unsigned long>(unsigned long)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::_Rb_tree_increment(std::_Rb_tree_node_base const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ios_base::~ios_base()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::range_error::~range_error()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__basic_file<char>::~__basic_file()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_guard_acquire@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream& std::ostream::_M_insert<bool>(bool)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::overflow_error@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_fstream<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::range_error@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_ios<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_filebuf<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `operator delete[](void*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(unsigned long, char, std::allocator<char> const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__detail::_List_node_base::_M_transfer(std::__detail::_List_node_base*, std::__detail::_List_node_base*)@GLIBCXX_3.4.15'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::replace(unsigned long, unsigned long, char const*, unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for std::exception@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::_Rep::_M_destroy(std::allocator<wchar_t> const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::istream& std::istream::_M_extract<double>(double&)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_filebuf<char, std::char_traits<char> >::close()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_fstream<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ifstream<char, std::char_traits<char> >::basic_ifstream(char const*, std::_Ios_Openmode)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::append(std::string const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `operator new(unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_istringstream<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for unsigned int@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::append(char const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::domain_error@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::find(char, unsigned long) const@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream::put(char)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for int@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__throw_bad_alloc()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_thread_atexit@CXXABI_1.3.7'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for unsigned int*@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::_Rb_tree_increment(std::_Rb_tree_node_base*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ifstream<char, std::char_traits<char> >::~basic_ifstream()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ios_base::Init::Init()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::condition_variable::condition_variable()@GLIBCXX_3.4.11'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_filebuf<char, std::char_traits<char> >::basic_filebuf()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::domain_error::~domain_error()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::cerr@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::find(char const*, unsigned long, unsigned long) const@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::allocator<char> const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::str() const@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::invalid_argument@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for void*@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::assign(std::string const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_ostringstream()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::_Rb_tree_rebalance_for_erase(std::_Rb_tree_node_base*, std::_Rb_tree_node_base&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for unsigned long@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__detail::_List_node_base::_M_hook(std::__detail::_List_node_base*)@GLIBCXX_3.4.15'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__detail::_List_node_base::_M_unhook()@GLIBCXX_3.4.15'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_ostringstream<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::_M_sync(char*, unsigned long, unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_iostream<char, std::char_traits<char> >::~basic_iostream()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::locale::locale(std::locale const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_istringstream<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `log2f@GLIBC_2.2.5'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream::operator<<(std::basic_streambuf<char, std::char_traits<char> >*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_streambuf<wchar_t, std::char_traits<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::exception::~exception()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__basic_file<char>::is_open() const@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_istringstream()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::swap(std::string&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_streambuf<char, std::char_traits<char> >::basic_streambuf(std::basic_streambuf<char, std::char_traits<char> > const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ios<char, std::char_traits<char> >::init(std::basic_streambuf<char, std::char_traits<char> >*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__throw_bad_cast()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ios<char, std::char_traits<char> >::clear(std::_Ios_Iostate)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_streambuf<wchar_t, std::char_traits<wchar_t> >::operator=(std::basic_streambuf<wchar_t, std::char_traits<wchar_t> > const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for long*@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `operator delete(void*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream::operator<<(int)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::_Rep::_S_empty_rep_storage@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::_Rep::_M_destroy(std::allocator<char> const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_iostream<wchar_t, std::char_traits<wchar_t> >::~basic_iostream()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::runtime_error@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_ofstream<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_stringstream()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `VTT for std::basic_stringstream<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream& std::ostream::_M_insert<long>(long)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::istream::get()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for unsigned long long@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::out_of_range::~out_of_range()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::length_error::~length_error()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::__ostream_insert<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*, long)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::invalid_argument::~invalid_argument()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::swap(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::cout@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream& std::ostream::_M_insert<unsigned long long>(unsigned long long)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream& std::ostream::_M_insert<void const*>(void const*)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::underflow_error@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_streambuf<char, std::char_traits<char> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for std::out_of_range@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_allocate_exception@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_ios<wchar_t, std::char_traits<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for void const*@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ios<wchar_t, std::char_traits<wchar_t> >::init(std::basic_streambuf<wchar_t, std::char_traits<wchar_t> >*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::reserve(unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_begin_catch@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for long@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::_Rep::_S_empty_rep_storage@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::_M_leak()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_filebuf<char, std::char_traits<char> >::open(char const*, std::_Ios_Openmode)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_stringbuf<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::_M_sync(wchar_t*, unsigned long, unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::istream::getline(char*, long, char)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_istream<char, std::char_traits<char> >& std::getline<char, std::char_traits<char>, std::allocator<char> >(std::basic_istream<char, std::char_traits<char> >&, std::basic_string<char, std::char_traits<char>, std::allocator<char> >&, char)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_stringstream<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::condition_variable::~condition_variable()@GLIBCXX_3.4.11'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::basic_stringbuf<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::insert(unsigned long, char const*, unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::string::assign(char const*, unsigned long)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for unsigned char@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ios_base::ios_base()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__throw_out_of_range(char const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::overflow_error::~overflow_error()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__throw_length_error(char const*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::__throw_system_error(int)@GLIBCXX_3.4.11'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ofstream<char, std::char_traits<char> >::close()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::ostream& std::ostream::_M_insert<double>(double)@GLIBCXX_3.4.9'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_streambuf<char, std::char_traits<char> >::operator=(std::basic_streambuf<char, std::char_traits<char> > const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `typeinfo for long long@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, unsigned long, std::allocator<char> const&)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_ifstream<char, std::char_traits<char> >::close()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_guard_release@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `__cxa_throw@CXXABI_1.3'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::underflow_error::~underflow_error()@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::_Rb_tree_decrement(std::_Rb_tree_node_base*)@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `vtable for std::length_error@GLIBCXX_3.4'\n",
      "/home/jyjiang/ls/envs/drug/compiler_compat/ld: /usr/local/cuda/lib64/libcufile.so: undefined reference to `std::basic_filebuf<char, std::char_traits<char> >::~basic_filebuf()@GLIBCXX_3.4'\n",
      "collect2: error: ld returned 1 exit status\n"
     ]
    }
   ],
   "source": [
    "from train_graph_davis import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "davis_drug_path = '/home/jyjiang/drug_interaction/data/drug.csv'\n",
    "davis_target_path = '/home/jyjiang/drug_interaction/data/bind_data/target.csv'\n",
    "davis_drug_feature_path = '/home/jyjiang/drug_interaction/data/bind_data/kpgt_base.npz'\n",
    "davis_protein_feature_path = '/home/jyjiang/drug_interaction/data/bind_data/prot_rep.pkl'\n",
    "davis_data_path = '/home/jyjiang/drug_interaction/data/bind_data/data.csv'\n",
    "davis_graph_folderpath = '/home/jyjiang/drug_interaction/data/bind_data/graph'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "davis_drug_feature = np.load(davis_drug_feature_path)\n",
    "davis_drug_feature = davis_drug_feature['fps']\n",
    "with  open(davis_protein_feature_path, 'rb') as f:\n",
    "    davis_protein_feature = pickle.load(f)\n",
    "# average to get same length\n",
    "davis_target_feature = []\n",
    "for feature in davis_protein_feature:\n",
    "    feature = feature.squeeze(0)\n",
    "    davis_target_feature.append(feature[1 : len(feature[0]) - 1].mean(0))\n",
    "# build dict of id and feature     \n",
    "davis_drug_df = pd.read_csv(davis_drug_path)\n",
    "davis_target_df = pd.read_csv(davis_target_path)\n",
    "davis_drug_ids = davis_drug_df['Drug_ID']\n",
    "davis_target_ids = davis_target_df['Target_ID']\n",
    "davis_drug_dict = dict(zip(davis_drug_ids, davis_drug_feature))\n",
    "# davis_target_dict = dict(zip(davis_target_ids, davis_target_feature))\n",
    "davis_data_df=pd.read_csv(davis_data_path)\n",
    "davis_data_df['Label']=davis_data_df['Label'].astype(int)\n",
    "# davis_drug_id_to_index = {drug_id: idx for idx, drug_id in enumerate(davis_drug_df['Drug_ID'])}\n",
    "# davis_target_id_to_index = {target_id: idx for idx, target_id in enumerate(davis_target_df['Target_ID'])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_dataset = GraphDataset_withsim(davis_data_df, davis_drug_df, davis_target_df, davis_drug_dict, davis_graph_folderpath)\n",
    "valid_dataset = GraphDataset_withsim(davis_data_df, davis_drug_df, davis_target_df, davis_drug_dict, davis_graph_folderpath)\n",
    "weight = torch.tensor([1.0, 2.0]).to(torch.device(f'cuda:0'))\n",
    "criterion = DTIContrastiveLoss()\n",
    "criterion2 = nn.CrossEntropyLoss(weight=weight)\n",
    "sim_loss_fn = SimilarityLoss()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_size_cam=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_loader = DataLoader(dataset=test_dataset, batch_size=batch_size_cam,collate_fn=custom_collate_fn_zero,shuffle=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GraphDTI(nn.Module):\n",
    "    def __init__(self, drug_dim, prot_dim, output_dim, perturb=False, surface_feature=False):\n",
    "        super(GraphDTI, self).__init__()\n",
    "        self.esm_linear = nn.Linear(prot_dim, 512)\n",
    "        self.esm_linear_sf = nn.Linear(prot_dim, 448)\n",
    "        # self.gcn = torch.nn.DataParallel(GraphCNN(pooling='MTP'), device_ids=[0, 1])\n",
    "        #self.gcn = GraphCNN(pooling='MTP')\n",
    "        self.gcn = GraphCNN(pooling='MTP')\n",
    "        self.surface_linear = nn.Linear(3, 16)\n",
    "        self.surface_linear2 = nn.Linear(16, 64)\n",
    "        \n",
    "        self.drug_linear = nn.Linear(drug_dim, 1024)\n",
    "        self.hidden2 = nn.Linear(1024, 256)\n",
    "        self.drug_prot_linear = nn.Linear(512, 256)\n",
    "        self.outlinear = nn.Linear(256, output_dim)\n",
    "        self.perturb = perturb\n",
    "        self.surface_feature = surface_feature\n",
    "        self.gradients = None\n",
    "    def activations_hook(self, grad):\n",
    "        self.gradients = grad\n",
    "    def get_features_gradients(self):\n",
    "        return self.gradients\n",
    "    def forward(self, drug, prot_data):\n",
    "        if self.surface_feature:\n",
    "            x_esm = self.esm_linear_sf(prot_data.x)\n",
    "            x_sf = self.surface_linear(prot_data.surface_feature)\n",
    "            x_sf = self.surface_linear2(x_sf)\n",
    "            x1 = torch.cat((x_esm, x_sf), dim=1)\n",
    "        else:\n",
    "            x1 = self.esm_linear(prot_data.x)\n",
    "            x1.register_hook(self.activations_hook)\n",
    "        gcn_n_feat1, gcn_g_feat1 = self.gcn(x1, prot_data)\n",
    "        if self.perturb:\n",
    "            gcn_n_feat2, gcn_g_feat2 = self.gcn(x1, prot_data, pertubed=True)\n",
    "            \n",
    "            x2 = self.drug_linear(drug)\n",
    "            x2 = F.relu(x2)\n",
    "            x2 = F.dropout(x2, 0.2)\n",
    "            x2 = self.hidden2(x2)\n",
    "            # x2 = F.relu(x2)\n",
    "            # x2 = F.dropout(x2, 0.2)\n",
    "            \n",
    "            x3 = torch.cat((x2, gcn_g_feat1), dim=1)\n",
    "            x3 = self.drug_prot_linear(x3)\n",
    "            x3 = F.relu(x3)\n",
    "            x3 = F.dropout(x3, 0.2)\n",
    "            out = self.outlinear(x3)\n",
    "            \n",
    "            return out, gcn_g_feat1, gcn_g_feat2\n",
    "        \n",
    "        else:\n",
    "            x2 = self.drug_linear(drug)\n",
    "            x2 = F.relu(x2)\n",
    "            x2 = F.dropout(x2, 0.2)\n",
    "            x2 = self.hidden2(x2)\n",
    "            # x2 = F.relu(x2)\n",
    "            # x2 = F.dropout(x2, 0.2)\n",
    "            \n",
    "            x3 = torch.cat((x2, gcn_g_feat1), dim=1)\n",
    "            x3 = self.drug_prot_linear(x3)\n",
    "            x3 = F.relu(x3)\n",
    "            x3 = F.dropout(x3, 0.2)\n",
    "            out = self.outlinear(x3)\n",
    "            return out,x2, gcn_g_feat1\n",
    "            #return out, x2, gcn_g_feat1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import PDB\n",
    "from collections import defaultdict\n",
    "\n",
    "def extract_residue_numbers(pdb_file):\n",
    "    \"\"\"Extract all residue numbers from a PDB file\"\"\"\n",
    "    parser = PDB.PDBParser(QUIET=True)\n",
    "    structure = parser.get_structure('protein', pdb_file)\n",
    "    \n",
    "    L=[]\n",
    "    for model in structure:\n",
    "        for chain in model:\n",
    "            residue_numbers = set()\n",
    "            for residue in chain:\n",
    "                if PDB.is_aa(residue):  \n",
    "                    resnum = residue.get_id()[1]\n",
    "                    residue_numbers.add(resnum)\n",
    "            L.append(sorted(residue_numbers))\n",
    "    return L\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bindpos(full_pdb,partial_pdb):\n",
    "    partial_residues = extract_residue_numbers(partial_pdb)\n",
    "    full_residues=extract_residue_numbers(full_pdb)\n",
    "    positions=[]\n",
    "    prev=0\n",
    "    for i in range(len(partial_residues)):\n",
    "        if partial_residues[i]:\n",
    "            positions+=[(res+prev) for res in partial_residues[i]] \n",
    "            prev+=max(full_residues[i])\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB import PDBParser, is_aa\n",
    "from Bio.SeqUtils import seq1\n",
    "\n",
    "def extract_sequence(pdbfile):\n",
    "    \"\"\"\n",
    "    Extract the sequence of amino acids from a PDB file using Biopython's PDBParser.\n",
    "    \"\"\"\n",
    "    parser = PDBParser(QUIET=True)\n",
    "    structure = parser.get_structure(pdbfile.split('/')[-1].split('.')[0], pdbfile)\n",
    "    \n",
    "    # Extract residues and convert to a sequence\n",
    "    residues = [residue for residue in structure.get_residues() if is_aa(residue, standard=True)]\n",
    "    sequence = \"\".join([seq1(residue.get_resname()) for residue in residues])  # Convert to one-letter code\n",
    "    \n",
    "    return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jyjiang/ls/envs/drug/lib/python3.12/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "def get_bindpos(full_pdb, partial_pdb):\n",
    "    \"\"\"\n",
    "    Compute the binding positions of the partial PDB file in the full PDB file\n",
    "    using sequence alignment.\n",
    "    \"\"\"\n",
    "    # Extract sequences from the PDB files\n",
    "    full_sequence = extract_sequence(full_pdb)\n",
    "    partial_sequence = extract_sequence(partial_pdb)\n",
    "    \n",
    "    # Perform sequence alignment\n",
    "    alignments = pairwise2.align.globalxx(full_sequence, partial_sequence)\n",
    "\n",
    "    # Use the best alignment (highest score)\n",
    "    best_alignment = alignments[0]\n",
    "    full_aligned, partial_aligned, score, start, end = best_alignment\n",
    "\n",
    "    # Map positions from the aligned sequences\n",
    "    positions = []\n",
    "    full_index = 0\n",
    "    partial_index = 0\n",
    "\n",
    "    for full_res, partial_res in zip(full_aligned, partial_aligned):\n",
    "        if full_res == partial_res and partial_res != \"-\":\n",
    "            positions.append(full_index + 1)  # Add 1 to convert to 1-based indexing\n",
    "        if full_res != \"-\":\n",
    "            full_index += 1\n",
    "        if partial_res != \"-\":\n",
    "            partial_index += 1\n",
    "\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "L=davis_data_df['Target_ID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "bindposmap=[]\n",
    "for pdb_id in L:\n",
    "    full_pdb=f'/home/jyjiang/drug_interaction/data/demo/{pdb_id}/{pdb_id}_protein.pdb'\n",
    "    partial_pdb=f'/home/jyjiang/drug_interaction/data/demo/{pdb_id}/{pdb_id}_pocket.pdb'\n",
    "    bindposmap.append( get_bindpos(full_pdb,partial_pdb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3807893/4036695014.py:2: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  model.load_state_dict(torch.load('/home/jyjiang/drug_interaction/best_model79.pth'))\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GraphDTI(\n",
       "  (esm_linear): Linear(in_features=1280, out_features=512, bias=True)\n",
       "  (esm_linear_sf): Linear(in_features=1280, out_features=448, bias=True)\n",
       "  (gcn): GraphCNN(\n",
       "    (gcn): ModuleList(\n",
       "      (0-2): 3 x GCNConv(512, 512)\n",
       "    )\n",
       "    (pool): GraphMultisetTransformer(512, 256, pool_sequences=['GMPool_G', 'GMPool_G'])\n",
       "    (drop1): Dropout(p=0.2, inplace=False)\n",
       "  )\n",
       "  (surface_linear): Linear(in_features=3, out_features=16, bias=True)\n",
       "  (surface_linear2): Linear(in_features=16, out_features=64, bias=True)\n",
       "  (drug_linear): Linear(in_features=2304, out_features=1024, bias=True)\n",
       "  (hidden2): Linear(in_features=1024, out_features=256, bias=True)\n",
       "  (drug_prot_linear): Linear(in_features=512, out_features=256, bias=True)\n",
       "  (outlinear): Linear(in_features=256, out_features=2, bias=True)\n",
       ")"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = GraphDTI(bindingdb_drug_feature[0].shape[0], bindingdb_target_feature[0].shape[0], 2).to(torch.device(f'cuda:0'))\n",
    "model.load_state_dict(torch.load('/home/jyjiang/drug_interaction/best_model79.pth'))\n",
    "model.eval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "252"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(test_loader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "matchness.append(np.sum( [1 if heatmap[x-1]>0 else 0 for x in bindposmap[i]])/len(bindposmap[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "matchness=[]\n",
    "maps=[]\n",
    "for i,b in enumerate(test_loader):\n",
    "    outputs,emb1, emb2 = model(b[0],b[1])\n",
    "    loss = criterion2(outputs.float(), b[2])\n",
    "    with torch.no_grad():\n",
    "      loss.backward()\n",
    "    pooled_gradients = torch.mean(model.gradients.squeeze(), dim=[1])\n",
    "    heatmap=pooled_gradients.cpu()\n",
    "    heatmap=np.maximum(0, heatmap)\n",
    "    maps.append(heatmap)\n",
    "    matchness.append(np.mean([heatmap[x-1] for x in bindposmap[i]])/torch.mean(heatmap) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def check_zero_percentage(tensor, per=0.3):\n",
    "    \"\"\"\n",
    "    Check if the percentage of zeros in the tensor is less than or equal to `per`.\n",
    "    \"\"\"\n",
    "    # Count the number of zeros in the tensor\n",
    "    zero_count = (tensor == 0).sum().item()\n",
    "    total_elements = tensor.numel()\n",
    "    \n",
    "    # Calculate the percentage of zeros\n",
    "    zero_percentage = zero_count / total_elements\n",
    "    \n",
    "    # Return whether the zero percentage is within the specified limit\n",
    "    return zero_percentage <= per\n",
    "\n",
    "def top_n_indices(lst, maps, n, per=0.3):\n",
    "    \"\"\"\n",
    "    Get the indices of the top N maximum values in `lst` filtered by zero percentage in `maps`.\n",
    "    \n",
    "    Parameters:\n",
    "    - lst: List of values (e.g., scores).\n",
    "    - maps: List of tensors corresponding to `lst`.\n",
    "    - n: Number of top indices to return.\n",
    "    - per: Maximum allowed percentage of zeros in the tensors in `maps`.\n",
    "    \"\"\"\n",
    "    # Convert the list to a NumPy array\n",
    "    arr = np.array(lst)\n",
    "    \n",
    "    # Filter out NaN values from the array\n",
    "    valid_indices = np.argwhere(~np.isnan(arr)).flatten()\n",
    "    valid_values = arr[valid_indices]\n",
    "    \n",
    "    # Filter indices based on the zero percentage in `maps`\n",
    "    filtered_indices = [\n",
    "        idx for idx in valid_indices \n",
    "        if check_zero_percentage(maps[idx], per=per)\n",
    "    ]\n",
    "    \n",
    "    # Sort the filtered indices by the values in `lst` (descending order)\n",
    "    sorted_indices = sorted(\n",
    "        filtered_indices, key=lambda idx: arr[idx], reverse=True\n",
    "    )\n",
    "    \n",
    "    # Return the top N indices (or fewer if there aren't enough valid indices)\n",
    "    return sorted_indices[:n]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[81, 146, 17, 222, 229]"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_n_indices(matchness, maps,5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1z6e', '1lpg', '4ddh', '1p1n', '2yki']"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[L[i] for i in top_n_indices(matchness,maps, 5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "select=[maps[i] for i in top_n_indices(matchness, maps,5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([8.8711e-05, 1.0924e-04, 8.6138e-05, 5.8467e-05, 3.5810e-05, 2.5157e-05,\n",
       "        2.1785e-05, 1.6240e-05, 1.3509e-05, 1.6405e-05, 1.1433e-05, 2.3069e-05,\n",
       "        2.2455e-05, 3.6293e-05, 4.3669e-05, 4.1387e-05, 3.4197e-05, 2.5251e-05,\n",
       "        1.7393e-05, 1.3475e-05, 2.4970e-06, 0.0000e+00, 0.0000e+00, 3.3077e-06,\n",
       "        1.2921e-05, 3.0715e-05, 3.7822e-05, 4.2307e-05, 4.2585e-05, 4.4579e-05,\n",
       "        3.3327e-05, 3.0045e-05, 1.4901e-05, 1.2059e-05, 1.0929e-05, 2.1361e-05,\n",
       "        2.4150e-05, 3.8419e-05, 4.3813e-05, 5.1407e-05, 3.7158e-05, 5.0221e-05,\n",
       "        4.0603e-05, 3.9604e-05, 4.1038e-05, 3.9646e-05, 3.2175e-05, 3.2851e-05,\n",
       "        2.3494e-05, 3.2634e-05, 3.5833e-05, 3.8479e-05, 4.2803e-05, 3.9381e-05,\n",
       "        3.2314e-05, 1.0178e-05, 2.3944e-07, 0.0000e+00, 0.0000e+00, 0.0000e+00,\n",
       "        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,\n",
       "        3.9319e-05, 4.0357e-05, 3.6364e-05, 3.3645e-05, 3.9074e-05, 3.2557e-05,\n",
       "        3.7088e-05, 3.1398e-05, 2.9715e-05, 1.6644e-05, 2.4568e-05, 0.0000e+00,\n",
       "        2.4957e-05, 4.7783e-05, 8.0195e-05, 8.2888e-05, 8.5853e-05, 7.7725e-05,\n",
       "        8.8939e-05, 6.9444e-05, 4.7749e-05, 5.5497e-05, 3.9723e-05, 3.0238e-05,\n",
       "        3.3499e-05, 3.5971e-05, 3.6788e-05, 2.9993e-05, 2.4524e-05, 1.4588e-05,\n",
       "        5.3788e-06, 6.9510e-06, 1.1708e-05, 2.1631e-05, 3.0919e-05, 2.9318e-05,\n",
       "        2.7194e-05, 2.9555e-05, 4.3615e-05, 4.6365e-05, 5.0446e-05, 5.0075e-05,\n",
       "        3.7707e-05, 6.2007e-05, 5.3569e-05, 6.0486e-05, 7.1539e-05, 9.1998e-05,\n",
       "        9.4239e-05, 9.2377e-05, 8.9751e-05, 1.0608e-04, 1.1151e-04, 1.2756e-04,\n",
       "        1.1860e-04, 1.3173e-04, 1.1455e-04, 1.3648e-04, 1.3687e-04, 9.5286e-05,\n",
       "        8.7347e-05, 6.6444e-05, 6.2201e-05, 5.7930e-05, 6.8875e-05, 8.5290e-05,\n",
       "        8.6001e-05, 7.9793e-05, 8.6561e-05, 5.6377e-05, 6.0914e-05, 5.8123e-05,\n",
       "        5.5298e-05, 4.8072e-05, 1.1493e-05, 1.7562e-05, 3.4774e-05, 4.1288e-05,\n",
       "        5.2149e-05, 9.1173e-05, 8.6010e-05, 1.3160e-04, 1.3950e-04, 1.5787e-04,\n",
       "        1.2474e-04, 8.4199e-05, 7.7975e-05, 6.3056e-05, 8.4576e-05, 1.1310e-04,\n",
       "        7.2848e-05, 7.2363e-05, 1.0872e-04, 7.8074e-05, 4.5010e-05, 4.6859e-05,\n",
       "        5.6919e-05, 5.7459e-05, 4.6140e-05, 5.0788e-05, 6.0782e-05, 7.1068e-05,\n",
       "        8.4079e-05, 1.1794e-04, 1.4508e-04, 1.6149e-04, 2.0346e-04, 1.9001e-04,\n",
       "        1.8631e-04, 1.6377e-04, 1.5850e-04, 1.4185e-04, 1.5240e-04, 1.1689e-04,\n",
       "        9.7672e-05, 7.3700e-05, 5.2576e-05, 6.9546e-05, 6.0583e-05, 4.9489e-05,\n",
       "        6.1023e-05, 7.2765e-05, 9.1093e-05, 8.5724e-05, 1.0576e-04, 7.0840e-05,\n",
       "        5.2371e-05, 2.7892e-05, 2.9494e-05, 3.9136e-05, 5.6117e-05, 5.9526e-05,\n",
       "        6.4979e-05, 9.5166e-05, 5.8572e-05, 6.4103e-05, 7.8747e-05, 7.1638e-05,\n",
       "        8.2836e-05, 9.7040e-05, 1.0295e-04, 1.0490e-04, 1.3851e-04, 1.9373e-04,\n",
       "        1.7223e-04, 1.7258e-04, 1.6559e-04, 1.8479e-04, 1.8380e-04, 1.2927e-04,\n",
       "        9.4140e-05, 8.7281e-05, 7.0379e-05, 6.2805e-05, 4.7344e-05, 6.8622e-05,\n",
       "        3.9637e-05, 2.5039e-05, 3.7008e-05, 1.8282e-05, 1.8916e-05, 2.1704e-05,\n",
       "        2.6366e-05, 2.8444e-05, 3.0720e-05, 2.8754e-05, 3.0870e-05, 3.4185e-05,\n",
       "        3.9103e-05, 5.7931e-05, 6.9859e-05, 5.3215e-05, 5.8519e-05, 6.6315e-05,\n",
       "        8.8145e-05, 6.8141e-05, 7.1670e-05, 4.6616e-05, 5.2384e-05, 4.8789e-05,\n",
       "        5.7435e-05, 5.0907e-05, 3.7138e-05, 4.9466e-05, 2.4400e-05, 2.4657e-05,\n",
       "        3.6378e-05, 3.0403e-05, 4.6327e-05, 1.7076e-05, 4.3370e-05, 3.7903e-05,\n",
       "        3.8513e-05, 4.0373e-05, 4.1402e-05, 4.6062e-05, 5.2793e-05, 4.1242e-05,\n",
       "        1.7567e-05, 1.2604e-05, 2.1360e-06, 2.1360e-06, 0.0000e+00, 1.9101e-05,\n",
       "        2.4086e-05, 3.6028e-05, 2.3100e-05, 4.2430e-05, 4.3760e-05, 4.4492e-05,\n",
       "        4.5690e-05, 4.7980e-05, 5.1778e-05, 5.7675e-05, 4.9125e-05, 4.8178e-05,\n",
       "        1.7402e-05, 3.0893e-05, 1.7749e-05, 2.5931e-05])"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "torch.save(select, \"select_heatmap.pt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6.555453642593658,\n",
       " 21.03174603174603,\n",
       " 0,\n",
       " 0.21031746031746032,\n",
       " 1.0,\n",
       " 0.21031746031746032,\n",
       " 0.3475409836065574,\n",
       " 0,\n",
       " 0,\n",
       " 0)"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evaluate(model, test_loader, criterion2,device='cuda')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jyjiang/drug_interaction/data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jyjiang/ls/envs/drug/lib/python3.12/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "%cd data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[11:26:20] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[11:26:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:21] Explicit valence for atom # 19 C, 5, is greater than permitted\n",
      "[11:26:21] ERROR: Could not sanitize molecule ending on line 78\n",
      "[11:26:21] ERROR: Explicit valence for atom # 19 C, 5, is greater than permitted\n",
      "[11:26:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:23] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:24] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:24] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:24] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:27] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:37] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:43] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:46] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:46] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:53] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:54] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:26:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:02] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:02] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:02] Explicit valence for atom # 3 N, 3, is greater than permitted\n",
      "[11:27:02] ERROR: Could not sanitize molecule ending on line 93\n",
      "[11:27:02] ERROR: Explicit valence for atom # 3 N, 3, is greater than permitted\n",
      "[11:27:02] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:02] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:03] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:03] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:03] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:04] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:07] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:09] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:09] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:10] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:10] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:10] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:11] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:11] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:13] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:15] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:15] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:15] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:16] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:16] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:17] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:17] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:18] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:19] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:19] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:19] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:20] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:20] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:20] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:25] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:25] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:27] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:28] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:28] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:31] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:31] Explicit valence for atom # 6 N, 4, is greater than permitted\n",
      "[11:27:31] ERROR: Could not sanitize molecule ending on line 143\n",
      "[11:27:31] ERROR: Explicit valence for atom # 6 N, 4, is greater than permitted\n",
      "[11:27:31] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:34] Explicit valence for atom # 13 C, 5, is greater than permitted\n",
      "[11:27:34] ERROR: Could not sanitize molecule ending on line 86\n",
      "[11:27:34] ERROR: Explicit valence for atom # 13 C, 5, is greater than permitted\n",
      "[11:27:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:34] Explicit valence for atom # 11 N, 5, is greater than permitted\n",
      "[11:27:34] ERROR: Could not sanitize molecule ending on line 104\n",
      "[11:27:34] ERROR: Explicit valence for atom # 11 N, 5, is greater than permitted\n",
      "[11:27:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:38] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:38] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:38] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:38] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:40] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:40] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:40] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:41] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:41] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:41] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:41] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:43] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:43] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:46] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:48] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:49] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:49] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:49] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:50] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:50] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:51] Explicit valence for atom # 7 N, 5, is greater than permitted\n",
      "[11:27:51] ERROR: Could not sanitize molecule ending on line 111\n",
      "[11:27:51] ERROR: Explicit valence for atom # 7 N, 5, is greater than permitted\n",
      "[11:27:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:51] Explicit valence for atom # 12 N, 5, is greater than permitted\n",
      "[11:27:51] ERROR: Could not sanitize molecule ending on line 110\n",
      "[11:27:51] ERROR: Explicit valence for atom # 12 N, 5, is greater than permitted\n",
      "[11:27:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:53] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:55] Explicit valence for atom # 1 O, 3, is greater than permitted\n",
      "[11:27:55] ERROR: Could not sanitize molecule ending on line 59\n",
      "[11:27:55] ERROR: Explicit valence for atom # 1 O, 3, is greater than permitted\n",
      "[11:27:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:56] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:56] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:56] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:56] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:56] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:56] Explicit valence for atom # 10 N, 5, is greater than permitted\n",
      "[11:27:56] ERROR: Could not sanitize molecule ending on line 107\n",
      "[11:27:56] ERROR: Explicit valence for atom # 10 N, 5, is greater than permitted\n",
      "[11:27:56] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:58] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:59] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:27:59] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:00] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:00] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:00] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:00] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:01] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:03] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:03] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:03] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:03] Explicit valence for atom # 4 N, 3, is greater than permitted\n",
      "[11:28:03] ERROR: Could not sanitize molecule ending on line 56\n",
      "[11:28:03] ERROR: Explicit valence for atom # 4 N, 3, is greater than permitted\n",
      "[11:28:03] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:04] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:04] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:04] Explicit valence for atom # 3 N, 4, is greater than permitted\n",
      "[11:28:04] ERROR: Could not sanitize molecule ending on line 53\n",
      "[11:28:04] ERROR: Explicit valence for atom # 3 N, 4, is greater than permitted\n",
      "[11:28:04] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:05] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:06] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:06] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:06] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:07] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:07] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:07] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:08] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:08] Explicit valence for atom # 4 N, 3, is greater than permitted\n",
      "[11:28:08] ERROR: Could not sanitize molecule ending on line 69\n",
      "[11:28:08] ERROR: Explicit valence for atom # 4 N, 3, is greater than permitted\n",
      "[11:28:08] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:09] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:09] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:11] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:11] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:11] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:12] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:14] Explicit valence for atom # 5 N, 4, is greater than permitted\n",
      "[11:28:14] ERROR: Could not sanitize molecule ending on line 104\n",
      "[11:28:14] ERROR: Explicit valence for atom # 5 N, 4, is greater than permitted\n",
      "[11:28:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:14] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:15] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:16] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:16] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:16] Explicit valence for atom # 13 N, 5, is greater than permitted\n",
      "[11:28:16] ERROR: Could not sanitize molecule ending on line 107\n",
      "[11:28:16] ERROR: Explicit valence for atom # 13 N, 5, is greater than permitted\n",
      "[11:28:16] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:16] Explicit valence for atom # 8 N, 5, is greater than permitted\n",
      "[11:28:16] ERROR: Could not sanitize molecule ending on line 71\n",
      "[11:28:16] ERROR: Explicit valence for atom # 8 N, 5, is greater than permitted\n",
      "[11:28:16] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:21] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:24] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:26] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:27] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:27] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:29] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:29] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:31] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:31] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:32] Explicit valence for atom # 14 C, 5, is greater than permitted\n",
      "[11:28:32] ERROR: Could not sanitize molecule ending on line 63\n",
      "[11:28:32] ERROR: Explicit valence for atom # 14 C, 5, is greater than permitted\n",
      "[11:28:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:32] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:33] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:34] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:35] Explicit valence for atom # 8 N, 3, is greater than permitted\n",
      "[11:28:35] ERROR: Could not sanitize molecule ending on line 129\n",
      "[11:28:35] ERROR: Explicit valence for atom # 8 N, 3, is greater than permitted\n",
      "[11:28:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:35] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:37] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:37] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:37] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:38] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:38] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:38] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:39] Explicit valence for atom # 2 N, 4, is greater than permitted\n",
      "[11:28:39] ERROR: Could not sanitize molecule ending on line 137\n",
      "[11:28:39] ERROR: Explicit valence for atom # 2 N, 4, is greater than permitted\n",
      "[11:28:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:39] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:41] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:41] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:42] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:44] Explicit valence for atom # 0 O, 3, is greater than permitted\n",
      "[11:28:44] ERROR: Could not sanitize molecule ending on line 52\n",
      "[11:28:44] ERROR: Explicit valence for atom # 0 O, 3, is greater than permitted\n",
      "[11:28:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:44] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:45] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:50] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:50] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:50] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:51] Explicit valence for atom # 11 N, 5, is greater than permitted\n",
      "[11:28:51] ERROR: Could not sanitize molecule ending on line 107\n",
      "[11:28:51] ERROR: Explicit valence for atom # 11 N, 5, is greater than permitted\n",
      "[11:28:51] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:52] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:54] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:54] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:54] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:55] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:28:57] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:29:00] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:29:00] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:29:00] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:29:01] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:29:01] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:29:01] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n",
      "[11:29:01] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CSV files created successfully!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import csv\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import AllChem\n",
    "from Bio import SeqIO\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "import numpy as np\n",
    "# Paths\n",
    "base_folder = \"/home/jyjiang/drug_interaction/data/demo\"  # Replace with your actual folder path\n",
    "output_data_csv = \"data.csv\"\n",
    "output_drug_csv = \"drug.csv\"\n",
    "output_target_csv = \"target.csv\"\n",
    "\n",
    "# Initialize containers for data\n",
    "data_rows = []\n",
    "drug_rows = {}\n",
    "target_rows = {}\n",
    "\n",
    "# Function to extract SMILES from SDF or MOL2\n",
    "def extract_smiles(file_path):\n",
    "    if file_path.endswith(\".sdf\"):\n",
    "        supplier = Chem.SDMolSupplier(file_path)\n",
    "        for mol in supplier:\n",
    "            if mol is not None:\n",
    "                return Chem.MolToSmiles(mol)\n",
    "    elif file_path.endswith(\".mol2\"):\n",
    "        from openbabel import pybel\n",
    "        mols = pybel.readfile(\"mol2\", file_path)\n",
    "        for mol in mols:\n",
    "            return mol.write(\"smi\").strip()\n",
    "    return None\n",
    "\n",
    "# Function to extract protein sequence from PDB\n",
    "def load_predicted_PDB3(pdbfile):\n",
    "    parser = PDBParser(QUIET=True)\n",
    "    structure = parser.get_structure(\"protein\", pdbfile)\n",
    "    \n",
    "    sequence = []\n",
    "    ca_coords = []\n",
    "\n",
    "    # \n",
    "    for model in structure:\n",
    "        for chain in model:\n",
    "            for residue in chain:\n",
    "                # \n",
    "                if residue.id[0] != \" \":\n",
    "                    continue\n",
    "                # \n",
    "                if \"CA\" in residue:\n",
    "                    sequence.append(residue.resname)  # \n",
    "                    ca_coords.append(residue[\"CA\"].coord)  # C \n",
    "\n",
    "    # \n",
    "    aa_map = {\n",
    "        'ALA': 'A', 'CYS': 'C', 'ASP': 'D', 'GLU': 'E', 'PHE': 'F',\n",
    "        'GLY': 'G', 'HIS': 'H', 'ILE': 'I', 'LYS': 'K', 'LEU': 'L',\n",
    "        'MET': 'M', 'ASN': 'N', 'PRO': 'P', 'GLN': 'Q', 'ARG': 'R',\n",
    "        'SER': 'S', 'THR': 'T', 'VAL': 'V', 'TRP': 'W', 'TYR': 'Y', 'SEC': 'U'\n",
    "    }\n",
    "    sequence = ''.join([aa_map.get(res, 'X') for res in sequence])  # \n",
    "    \n",
    "    \n",
    "    num_atoms = np.array(ca_coords).shape[0]\n",
    "    distance_matrix = np.zeros((num_atoms, num_atoms))\n",
    "    \n",
    "    for i in range(num_atoms):\n",
    "        for j in range(i, num_atoms):\n",
    "            distance = np.linalg.norm(ca_coords[i] - ca_coords[j])  # \n",
    "            distance_matrix[i, j] = distance\n",
    "            distance_matrix[j, i] = distance  # \n",
    "\n",
    "    return distance_matrix, sequence\n",
    "\n",
    "# Traverse the demo folder\n",
    "for subdir, _, files in os.walk(base_folder):\n",
    "    protein_id = os.path.basename(subdir)\n",
    "    ligand_sdf = os.path.join(subdir, f\"{protein_id}_ligand.sdf\")\n",
    "    ligand_mol2 = os.path.join(subdir, f\"{protein_id}_ligand.mol2\")\n",
    "    protein_pdb = os.path.join(subdir, f\"{protein_id}_protein.pdb\")\n",
    "\n",
    "    # Skip if required files are missing\n",
    "    if not os.path.exists(protein_pdb) or (not os.path.exists(ligand_sdf) and not os.path.exists(ligand_mol2)):\n",
    "        continue\n",
    "\n",
    "    # Extract SMILES\n",
    "    smiles = extract_smiles(ligand_sdf if os.path.exists(ligand_sdf) else ligand_mol2)\n",
    "    if not smiles:\n",
    "        continue\n",
    "\n",
    "    # Extract protein sequence\n",
    "    distance_matrix,protein_seq = load_predicted_PDB3(protein_pdb)\n",
    "    if not protein_seq or len(protein_seq>1200):\n",
    "        continue\n",
    "\n",
    "    # Create unique IDs for the drug and target\n",
    "    drug_id = f\"{len(drug_rows) + 1}\"\n",
    "    target_id = protein_id\n",
    "\n",
    "    # Add to drug and target rows\n",
    "    if smiles not in drug_rows:\n",
    "        drug_rows[smiles] = drug_id\n",
    "    if target_id not in target_rows:\n",
    "        target_rows[target_id] = protein_seq\n",
    "\n",
    "    # Add to data rows\n",
    "    data_rows.append({\n",
    "        \"Drug_ID\": drug_rows[smiles],\n",
    "        \"Drug\": smiles,\n",
    "        \"Target_ID\": target_id,\n",
    "        \"Target\": protein_seq,\n",
    "        \"Label\": 1  # Default label (can be adjusted based on your needs)\n",
    "    })\n",
    "\n",
    "# Write data.csv\n",
    "with open(output_data_csv, \"w\", newline=\"\") as csvfile:\n",
    "    fieldnames = [\"Drug_ID\", \"Drug\", \"Target_ID\", \"Target\", \"Label\"]\n",
    "    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)\n",
    "    writer.writeheader()\n",
    "    writer.writerows(data_rows)\n",
    "\n",
    "# Write drug.csv\n",
    "with open(output_drug_csv, \"w\", newline=\"\") as csvfile:\n",
    "    fieldnames = [\"Drug_ID\", \"smiles\"]\n",
    "    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)\n",
    "    writer.writeheader()\n",
    "    for smiles, drug_id in drug_rows.items():\n",
    "        writer.writerow({\"Drug_ID\": drug_id, \"smiles\": smiles})\n",
    "\n",
    "# Write target.csv\n",
    "with open(output_target_csv, \"w\", newline=\"\") as csvfile:\n",
    "    fieldnames = [\"Target_ID\", \"Target\"]\n",
    "    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)\n",
    "    writer.writeheader()\n",
    "    for target_id, protein_seq in target_rows.items():\n",
    "        writer.writerow({\"Target_ID\": target_id, \"Target\": protein_seq})\n",
    "\n",
    "print(\"CSV files created successfully!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "drug",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
